src/python/grpcio/grpc/_channel.py

```diff
@@ -314,70 +363,111 @@ def add_callback(self, callback):
 
     def initial_metadata(self):
         """See grpc.Call.initial_metadata"""
+        raise NotImplementedError()
+
+    def trailing_metadata(self):
+        """See grpc.Call.trailing_metadata"""
+        raise NotImplementedError()
+
+    def code(self):
+        """See grpc.Call.code"""
+        raise NotImplementedError()
+
+    def details(self):
+        """See grpc.Call.details"""
+        raise NotImplementedError()
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        return self._next()
+
+    def __next__(self):
+        return self._next()
+
+    def _next(self):
+        raise NotImplementedError()
+
+    def debug_error_string(self):
+        raise NotImplementedError()
+
+    def _repr(self):
+        return _rpc_state_string(self.__class__.__name__, self._state)
+
+    def __repr__(self):
+        return self._repr()
+
+    def __str__(self):
+        return self._repr()
+
+    def __del__(self):
         with self._state.condition:
+            if self._state.code is None:
+                self._state.code = grpc.StatusCode.CANCELLED
+                self._state.details = 'Cancelled upon garbage collection!'
+                self._state.cancelled = True
+                self._call.cancel(
+                    _common.STATUS_CODE_TO_CYGRPC_STATUS_CODE[self._state.code],
+                    self._state.details)
+                self._state.condition.notify_all()
 
-            def _done():
-                return self._state.initial_metadata is not None
 
-            _common.wait(self._state.condition.wait, _done)
+class _SingleThreadedRendezvous(_Rendezvous):  # pylint: disable=too-many-ancestors
+    """An RPC iterator operating entirely on a single thread.
+
+    The __next__ method of _SingleThreadedRendezvous does not depend on the
+    existence of any other thread, including the "channel spin thread".
+    However, this means that its interface is entirely synchronous. So this
+    class cannot fulfill the grpc.Future interface.
+    """
+
+    def initial_metadata(self):
+        """See grpc.Call.initial_metadata"""
+        with self._state.condition:
+            # NOTE(gnossen): Based on our initial call batch, we are guaranteed
+            # to receive initial metadata before any messages.
+            while self._state.initial_metadata is None:
+                self._get_next_event()
             return self._state.initial_metadata
 
     def trailing_metadata(self):
         """See grpc.Call.trailing_metadata"""
         with self._state.condition:
-
-            def _done():
-                return self._state.trailing_metadata is not None
-
-            _common.wait(self._state.condition.wait, _done)
+            if self._state.trailing_metadata is None:
+                raise RuntimeError(
+                    "Cannot get trailing metadata until RPC is completed.")
```

Comment:
I don't think `grpc.RpcError` is semantically appropriate. I was thinking about adding a `grpc.UsageError` or something like that. The difference is that `grpc.RpcError` implies "something went wrong with the RPC". Whereas, these new exceptions mean "You're using the API wrong." But I was apprehensive because that means a new class in `__init__.py`. If you think it's worth it, I'll add it.

Reply:
I don't have strong opinion here, `RuntimeError` is also acceptable since we can't find a better type for it.

Ideally we can use [`concurrent.futures.InvalidStateError`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.InvalidStateError), but I just found out it is 3.8+.

About adding `grpc.UsageError`, I think it is optional, since this is for experimental path, it would make sense to add the new exception type to experimental folder first.

Reply:
Discussed offline. Adding `grpc.experimental.UsageError`.